xen/arch/arm/cpufreq/scpi_cpufreq.c

```diff
@@ -0,0 +1,353 @@
+/*
+ * xen/arch/arm/cpufreq/scpi_cpufreq.c
+ *
+ * SCPI based CPUFreq driver
+ *
+ * Based on Xen arch/x86/acpi/cpufreq/cpufreq.c
+ *
+ * Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
+ * Copyright (c) 2017 EPAM Systems.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <xen/types.h>
+#include <xen/delay.h>
+#include <xen/cpumask.h>
+#include <xen/sched.h>
+#include <xen/xmalloc.h>
+#include <xen/err.h>
+#include <xen/cpufreq.h>
+#include <asm/bug.h>
+#include <asm/percpu.h>
+
+#include "scpi_protocol.h"
+
+extern struct device *get_cpu_device(unsigned int cpu);
+
+struct scpi_cpufreq_data
+{
+    struct processor_performance *perf;
+    struct cpufreq_frequency_table *freq_table;
+    struct scpi_dvfs_info *info;
+    int domain_id;
+};
+
+static struct scpi_cpufreq_data *cpufreq_driver_data[NR_CPUS];
+
+static struct cpufreq_driver scpi_cpufreq_driver;
+
+static struct scpi_ops *scpi_ops;
+
+static unsigned int scpi_cpufreq_get(unsigned int cpu)
+{
+    struct scpi_cpufreq_data *data;
+    struct cpufreq_policy *policy;
+    const struct scpi_opp *opp;
+    int idx;
+
+    if ( cpu >= nr_cpu_ids || !cpu_online(cpu) )
+        return 0;
+
+    policy = per_cpu(cpufreq_cpu_policy, cpu);
+    if ( !policy || !( data = cpufreq_driver_data[policy->cpu] ) )
+        return 0;
+
+    /* Sanity check */
+    if ( data->domain_id < 0 || !data->info )
+        return 0;
+
+    idx = scpi_ops->dvfs_get_idx(data->domain_id);
+    if ( idx < 0 )
+        return 0;
+
+    opp = data->info->opps + idx;
+
+    return opp->freq / 1000; /* kHz */
+}
+
+static unsigned int check_freq(unsigned int cpu, unsigned int freq)
+{
+    unsigned int curr_freq;
+    int i;
+
+    for ( i = 0; i < 100; i++ )
```

Comment:
Why 100?
And why you need this loop at all? I think, comment is needed there.

Reply:
Well, I have borrowed this code from x86 CPUFreq, slightly modified it, but left time constants as is for now.
This function purpose is to check if a target frequency has already set up. Actually, I think that such code is *only needed* if management processor changes frequency in an asynchronous way. With SCP in ARM TF it is hard to implement. So, in order to be generic, I tend to leave this code for now, but going to reuse x86 acpi_pstate_strict property not to let this code execute by default. How does it sound?

Reply:
This functional was taken consciously) I just did not have time to reuse acpi_pstate_strict to disable such functionality by default, I even left todo regarding this. If you have some thoughts regarding how long we have to wait until the frequency is set up, please, propose your values.

Reply:
We decided to drop this func.

Reply:
E.g. transition latency?

Reply:
the whole check_freq().